slack/management/commands/fetch_slack_activity.py

```diff
@@ -0,0 +1,314 @@
+    selected_channels = []
+    if channels:
+        for channel_data in get_my_channels():
+            if channel_data["name"] in channels:
+                selected_channels.append(channel_data)
+                channels.remove(channel_data["name"])
+        if channels:
+            raise click.BadParameter(
+                "Could not find channels %r (maybe the bot isn't a member?)" % channels
+            )
+    else:
+        # materialize this generator so we can iterate multiple times
+        selected_channels.extend(get_my_channels())
+
+    for channel_data in selected_channels:
+        with transaction.atomic():
+            channel, created = Channel.objects.update_or_create(
+                id=channel_data["id"],
+                defaults={
+                    "name": channel_data["name"],
+                    "topic": channel_data["topic"]["value"],
+                    "purpose": channel_data["purpose"]["value"],
+                },
+            )
+            if created:
+                # we don't have any messages for this channel we just created
+                channel.channelupdategap_set.create(
+                    oldest_message_ts=None,
+                    newest_message_ts=None,
+                )
+            elif (
+                channel.last_update_ts
+                and not channel.channelupdategap_set.filter(
+                    newest_message_ts=None
+                ).exists()
+            ):
+                # gap from the most recent fetch till now
+                channel.channelupdategap_set.create(
+                    oldest_message_ts=channel.last_update_ts,
+                    newest_message_ts=None,
+                )
+            else:
+                assert (
+                    channel.channelupdategap_set.exists()
+                ), "We must have SOME gaps, time has passed since the last run!"
+
+    gaps = ChannelUpdateGap.objects.filter(
+        channel__id__in={c["id"] for c in selected_channels}
+    )
+    for gap in gaps:
+        fill_channel_gap(gap, debug)
+
+    # We have to track threads we've seen and update independently, replies
+    # don't show up in main channel history[1].
+    #
+    # [1]: <https://github.com/slackapi/python-slack-sdk/issues/1306>
+    logger.info("Fetching threads")
+    threads = (
+        Thread.objects.filter(channel_id__in={c["id"] for c in selected_channels})
+        .extra(
+            where=[
+                # Assume threads not updated for more than 1 month ago won't
+                # get posted to. Otherwise it's too much work to check all
+                # threads ever. last_update will be null for the threads
+                # do_channel just created, indicating they need to be updated
+                # at least once.
+                """
+                last_update_ts IS NULL
+                OR to_timestamp(last_update_ts::numeric) >= now() - INTERVAL '30 days'
+                """
+            ]
+        )
```

Comment:
We should really avoid using `extra`. I don't see any reason why we can't use the ORM for this. Off the top of my head, something like this should work (I've included the `CharField` change I suggested elsewhere):

```suggestion
        .filter(
            Q(last_update_ts="") | Q(last_update_ts__gte=one_month_ago)
        )
```

Where you calculate one month ago in python before this line.

Reply:
There's nothing wrong with extra().  Your suggestion uses lexicographic instead of the correct chronological ordering.

Reply:
> There's nothing wrong with extra()

The [Django documentation](https://docs.djangoproject.com/en/5.1/ref/models/querysets/#extra) recommends against using it except "as a last resort [...] only if you cannot express your query using other queryset methods", and goes so far as to request you file a ticket in core Django if you find you need to use it.

> Your suggestion uses lexicographic instead of the correct chronological ordering.

You're right - you'd need to create a custom db function class

```python
class ToTimestamp(Func):
    function = 'to_timestamp'
    output_field = DateTimeField()
```

and then do something like this (untested):

```python
.annotate(
    last_update_ts_converted=ToTimestamp(F('last_update_ts', output_field=DecimalField()))
).filter(
    Q(last_update_ts__isnull=True) |
    Q(last_update_ts_converted__gte=Now() - timedelta(days=30))
)
```

If you find this too annoying. You can use `RawSQL`, which is preferred over `extra`. I believe the syntax would be:

```python
.filter(
    Q(last_update_ts__isnull=True) |
    Q(RawSQL("to_timestamp(last_update_ts::numeric) >= now() - INTERVAL '30 days'", [])
)